### PR TITLE
improved handling of options in the Bash completion

### PR DESCRIPTION
**Pull request by carzacc**
This time the branches look right

### CODE REVIEW

**Comment by carzacc** [comment]
I still want to address the references I make to the third word, that the script assumes is the place where it can find the first option for the subcommand: in this new and more flexible script it just doesn't cut it anymore to assume that, it made sense to me at the time as a way to introduce flexibility but it's not enough really.

**Comment by carzacc** [comment]
Now I would be quite happy to merge this PR already, just leaving a note: for commands like filesystem destroy where the user can pick more than one value from a list I would like not to suggest to the user again stuff they've already picked, but I think we can leave that to a future PR.

**Comment by bgurney-rh** [comment]
I would be careful about auto-completing items after a "destroy" sub-command, to ensure that the user is making a conscious decision to destroy something they intend to destroy (hopefully after they have confirmed, via the output of "stratis pool list" or "stratis fs list", the item they wish to destroy).

**Comment by carzacc** [comment]
> I would be careful about auto-completing items after a "destroy" sub-command, to ensure that the user is making a conscious decision to destroy something they intend to destroy (hopefully after they have confirmed, via the output of "stratis pool list" or "stratis fs list", the item they wish to destroy).

The issue I saw (which is the one I mentioned in my comment) is that the user will receive autocomplete suggestions for filesystem "filesystem-one" again when they type `stratis fs destroy pool-name filesystem-one fil` and they hit Tab. 

To me, it seems you're suggesting the script shouldn't autocomplete filesystem names at all as a way to encourage user to use commands like `fs list` and do more thorough checking before issuing `fs destroy`.

It would be a fair point, and I agree that autocomplete makes it more likely that users will make this kind of choice lightly.

I don't think, though, that removing features from autocomplete, and making it less convenient to use (for example because of having to type long names) for everyone, just to encourage users to perform more thorough checks and know the exact filesystem name before issuing destructive commands is the best choice. Another thing to consider is that, to delete filesystems, there are two choices to make: first the user has to choose a pool and then a filesystem from that pool, which already narrows down the choice in some cases.

An alternative might be to add a prompt to the CLI asking users to confirm destructive action, perhaps including other data present in `fs list` like the creation date, the path and size which might cause a user to doubt what they're doing.

To avoid any potential confusion, I'm just trying contribute my own personal view to this discussion, if you tell me to remove all autocompletion for `destroy` subcommands I'll do it ASAP.

I hope my phrasing was clear.

**Comment by bgurney-rh** [comment]
Actually, after testing it, I think it's okay as is, and I was probably just being a little too cautious.

After some preliminary testing, the auto-complete helped me find the right sequence to set up an encrypted pool, after having known the general steps to do so, but forgetting the exact steps.  I think that's a pretty good result, considering that completion is supposed to help the user who is in this current state.

**Comment by bgurney-rh** [comment]
There's a few occurrences of trailing spaces that I'll highlight in a few minutes (including an extra line at the end of `shell-completion/bash/stratis` that was added).

File: shell-completion/bash/stratis [hunk 1]
```diff
@@ -1,70 +1,116 @@
...
+				fi
+			else
+				if [[ ${root_subcommand} != "" && ${subcommand} != "" && ${subcommand_opt} == "" ]]; then
+					subcommand_opt=${el}
+				fi
+				let "comp_cword_adj--"
+			fi
+
+		done
+		# find the subcommand
```
Comment (bgurney-rh) [inline shell-completion/bash/stratis:51]: There is a space at the end of `# find the subcommand`, and there are 4 tabs on the empty line on line 55, just before `break`.

File: shell-completion/bash/stratis [hunk 2]
```diff
@@ -1,70 +1,116 @@
...
+		case ${root_subcommand} in
+			pool)
+				if [[ ${subcommand} == "create" ]]; then
+					COMPREPLY=( $(compgen -W "${pool_create_opts} ${opts}" -- ${cur}) )
+				elif [[ ${subcommand} == "bind" ]] && [[ ${subcommand_first_param} != "tpm2" ]]; then
+					COMPREPLY=( $(compgen -W "${pool_bind_opts} ${opts}" -- ${cur}) )
+				else
 					COMPREPLY=( $(compgen -W "${opts}" -- ${cur}) )
+				fi
+
```
Comment (bgurney-rh) [inline shell-completion/bash/stratis:85]: There are four tabs on the empty line of line 85

File: shell-completion/bash/stratis [hunk 3]
```diff
@@ -1,70 +1,116 @@
...
 	daemon_subcommands="redundancy version"
 	key_subcommands="set unset reset list"
 	key_set_reset_opts="--keyfile-path --capture-key"
 	blockdevs="$(echo /dev/disk/*/*) $(lsblk -n -o name -p -l)"
+	root_subcommand=""
+	subcommand=""
+	subcommand_first_param=""
+	subcommand_opt=""
+	comp_cword_adj=${COMP_CWORD}
+
```
Comment (bgurney-rh) [inline shell-completion/bash/stratis:29]: There is one tab on the empty line on line 29.

**Comment by carzacc** [comment]
> There's a few occurrences of trailing spaces that I'll highlight in a few minutes (including an extra line at the end of `shell-completion/bash/stratis` that was added).

Done! If it looks good, I can squash the commits.

**Comment by bgurney-rh** [comment]
I think it looks good; go ahead and squash.

**Comment by mulkieran** [comment]
@carzacc Plz rebase and I'll merge this PR.

**Comment by carzacc** [comment]
> @carzacc Plz rebase and I'll merge this PR.

Is it good to merge now?

**Comment by mulkieran** [comment]
Yes, thx.

**Pull request merged**